broker/src/broker_server.cpp

```diff
@@ -0,0 +1,525 @@
+                                                   ieee1905_1::cCmduHeader::get_initial_size());
+    ieee1905_1::sTlvHeader *first_tlv = tlv;
+    message::sUdsHeader *uds_header   = static_cast<message::sUdsHeader *>(cmdu);
+
+    bool ret = true;
+
+    uint8_t type    = tlv->type;
+    uint16_t length = tlv->length;
+
+    do {
+
+        swap_16(length);
+        if (static_cast<ieee1905_1::eTlvType>(type) == ieee1905_1::eTlvType::TLV_VENDOR_SPECIFIC) {
+            auto tlv_vendor_specific = ieee1905_1::tlvVendorSpecific(
+                (uint8_t *)tlv, length + sizeof(ieee1905_1::sTlvHeader), true);
+            if (!tlv_vendor_specific.isInitialized()) {
+                LOG(ERROR) << "tlvVendorSpecific init() failure";
+                ret = false;
+                break;
+            }
+
+            if (tlv_vendor_specific.vendor_oui() ==
+                ieee1905_1::tlvVendorSpecific::eVendorOUI::OUI_INTEL) {
+                // assuming that the magic is the first data on the beerocks header
+                auto beerocks_magic = *(uint32_t *)(tlv_vendor_specific.payload());
+                swap_32(beerocks_magic);
+                if (beerocks_magic != message::MESSAGE_MAGIC) {
+                    LOG(WARNING) << "mismatch magic " << std::hex << int(beerocks_magic)
+                                 << " != " << int(message::MESSAGE_MAGIC) << std::dec;
+                    ret = false;
+                    break;
+                }
+
+            } else {
+                LOG(INFO) << "Not an Intel vendor specific message!";
+            }
+
+            // cancel the swap we did
+            tlv_vendor_specific.class_swap();
+        } else if (static_cast<ieee1905_1::eTlvType>(type) ==
+                       ieee1905_1::eTlvType::TLV_END_OF_MESSAGE &&
+                   length == 0) {
+            return true;
+        }
+
+        // set the next tlv
+        tlv  = (ieee1905_1::sTlvHeader *)((uint8_t *)tlv + sizeof(ieee1905_1::sTlvHeader) + length);
+        type = tlv->type;
+        length = tlv->length;
+
+    } while ((uint8_t *)tlv <= (uint8_t *)first_tlv + uds_header->length);
+
+    LOG(ERROR) << "TLV end of message not found! tlv_type=" << int(tlv->type)
+               << ", tlv_length=" << int(tlv->length); // << ", " << print_cmdu_types(uds_header);
+
+    std::ptrdiff_t available_bytes = uds_header->length + sizeof(message::sUdsHeader);
+    LOG(DEBUG) << "hex_dump (" + std::to_string(available_bytes) + " bytes):" << std::endl
+               << utils::dump_buffer(static_cast<const uint8_t *>(cmdu), available_bytes);
+
+    return ret;
+}
+
+static bool is_restricted_type(uint32_t type)
+{
+    // TODO: Until VS message filtering is implemented (based on action and action_op)
+    //       assume there are no restricted types. After the implementation,
+    //       subscribing to BROKER specific message should be restricted.
+
```

Comment:
Since filtering is performed on the broker side, I thought we could also filter vendor-specific messages (instead of forwarding to all the clients that subscribed to `ieee1905_1::eMessageType::VENDOR_SPECIFIC_MESSAGE`)